lib/ecco/row_type.rb

```diff
@@ -0,0 +1,32 @@
+module Ecco
+  class RowType
+    java_import com.github.shyiko.mysql.binlog.event.EventType
```

Comment:
since this import was added here I think it can be removed in `event_listener.rb`:

```diff
diff --git a/lib/ecco/event_listener.rb b/lib/ecco/event_listener.rb
index a07bf4a..7efb318 100644
--- a/lib/ecco/event_listener.rb
+++ b/lib/ecco/event_listener.rb
@@ -1,7 +1,6 @@
 module Ecco
   class EventListener
     include com.github.shyiko.mysql.binlog.BinaryLogClient::EventListener
-    java_import com.github.shyiko.mysql.binlog.event.EventType
 
     attr_writer :callback
 ```

Reply:
However, one can reason a bit whether some things that were moved to a completely new class (RowType) could be moved into the already existing parent class.

Reply:
Ah, so skip RowType entirely and move in the two classes in RowType into EventListener?

Reply:
Yeah, I think that could work out

Reply:
So something like this? 67bfd8fcbaa3c53d731ab7527b6a4f5b5d59135b

Reply:
Kind of, I don't think we need the inner class, also now we can work with instances instead of classes which I think could make things a bit more simpler even.

Reply:
Are you suggesting that I could move `EVENTS` into an instance variable `@events` instead and then in `SaveEventListener` override it to fit the latter instead?

Reply:
I can show what I mean here: 4b5f22cf3ac874f8642eea085e4c30be31db6d52

Reply:
> Are you suggesting that I could move `EVENTS` into an instance variable `@events` instead and then in `SaveEventListener` override it to fit the latter instead?

Almost, I think it's more common to work with overloading of methods than variables though

Reply:
Aha!

Reply:
I can try to make a diff tomorrow to show how it could look. Getting late now :)

Reply:
So something like this? e9551dfbab7cf682884ba33e53e32aac9d1c713a

Reply:
> I can try to make a diff tomorrow to show how it could look. Getting late now :)

No worries! No need to look at the latest commit :D

Reply:
I don't know if this is readable, but here's the full diff of an slightly alternative way of doing it. Basically I built on the ideas here, but I re-instated the old arrays to make it more clear in the subclasses what kind of events they care about. This also has the upshot of not generating the arrays on every event. Also changed the name of the helper to something more fitting (IMHO):

```diff
diff --git a/lib/ecco/event_listener.rb b/lib/ecco/event_listener.rb
index a41c8d3..b1d5a12 100644
--- a/lib/ecco/event_listener.rb
+++ b/lib/ecco/event_listener.rb
@@ -3,36 +3,32 @@ module Ecco
     include com.github.shyiko.mysql.binlog.BinaryLogClient::EventListener
     java_import com.github.shyiko.mysql.binlog.event.EventType
 
+    WRITE_EVENTS  = { EventType::WRITE_ROWS => "WRITE_ROWS", EventType::EXT_WRITE_ROWS => "WRITE_ROWS" }
+    UPDATE_EVENTS = { EventType::UPDATE_ROWS => "UPDATE_ROWS", EventType::EXT_UPDATE_ROWS => "UPDATE_ROWS" }
+    DELETE_EVENTS = { EventType::DELETE_ROWS => "DELETE_ROWS" , EventType::EXT_DELETE_ROWS => "DELETE_ROWS" }
+    QUERY_EVENTS  = { EventType::QUERY => "QUERY"}
+    ROTATE_EVENTS = { EventType::ROTATE => "ROTATE"}
+
+    TABLE_EVENT  = EventType::TABLE_MAP
+
     attr_writer :callback
 
-    def initialize(client)
+    def initialize(client, events_of_interest)
       @client = client
+      @events_of_interest = events_of_interest
       @callback = Proc.new {}
     end
 
-    TABLE_EVENT = EventType::TABLE_MAP
-
     def on_event(event)
       raise NotImplementedError
     end
 
-    def events
-      {
-        EventType::WRITE_ROWS      => "WRITE_ROWS",
-        EventType::EXT_WRITE_ROWS  => "WRITE_ROWS",
-        EventType::UPDATE_ROWS     => "UPDATE_ROWS",
-        EventType::EXT_UPDATE_ROWS => "UPDATE_ROWS",
-        EventType::DELETE_ROWS     => "DELETE_ROWS",
-        EventType::EXT_DELETE_ROWS => "DELETE_ROWS"
-      }
-    end
-
     def accepted_events
-      events.keys
+      @events_of_interest.keys
     end
 
-    def fetch_event(type)
-      events.fetch(type)
+    def event_type_to_string(type)
+      @events_of_interest.fetch(type)
     end
   end
 end
diff --git a/lib/ecco/row_event_listener.rb b/lib/ecco/row_event_listener.rb
index 5d78eb9..5dcc2fe 100644
--- a/lib/ecco/row_event_listener.rb
+++ b/lib/ecco/row_event_listener.rb
@@ -3,6 +3,12 @@ require "ecco/row_event"
 
 module Ecco
   class RowEventListener < EventListener
+    def initialize(client)
+      events_of_interest =
+        WRITE_EVENTS.merge(UPDATE_EVENTS).merge(DELETE_EVENTS)
+      super(client, events_of_interest)
+    end
+
     def on_event(event)
       data = event.get_data
       type = event.get_header.get_event_type
@@ -14,7 +20,7 @@ module Ecco
         row_event          = Ecco::RowEvent.new
         row_event.table_id = data.get_table_id
         row_event.rows     = data.rows
-        row_event.type     = fetch_event(type)
+        row_event.type     = event_type_to_string(type)
 
         if @table_map_event
           table_event_data = @table_map_event.get_data
diff --git a/lib/ecco/save_event_listener.rb b/lib/ecco/save_event_listener.rb
index 476aa2f..41fd52c 100644
--- a/lib/ecco/save_event_listener.rb
+++ b/lib/ecco/save_event_listener.rb
@@ -2,22 +2,22 @@ require "ecco/event_listener"
 
 module Ecco
   class SaveEventListener < EventListener
+    def initialize(client)
+      events_of_interest =
+        WRITE_EVENTS.merge(UPDATE_EVENTS).merge(DELETE_EVENTS).merge(QUERY_EVENTS).merge(ROTATE_EVENTS)
+      super(client, events_of_interest)
+    end
+
     def on_event(event)
       type = event.get_header.get_event_type
       case type
       when *accepted_events
         filename        = @client.get_binlog_filename
         position        = @client.get_binlog_position
-        event_type_name = fetch_event(type)
+        event_type_name = event_type_to_string(type)
 
         @callback.call(filename, position, event_type_name)
       end
     end
-    def events
-      {
-        EventType::QUERY  => "QUERY",
-        EventType::ROTATE => "ROTATE"
-      }.merge(super)
-    end
   end
 end
```

Reply:
I am not super-happy about the merging going on, there are probably better ways to solve that. Also the value of `#keys` should probably be stashed for some improved performance